File: src/common/meta/traces.rs
// Copyright 2025 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

use std::collections::HashMap;

use config::utils::json;
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Span {
    pub trace_id: String,
    pub span_id: String,
    pub flags: u8,
    pub span_status: String,
    pub span_kind: String,
    pub operation_name: String,
    pub start_time: u64,
    pub end_time: u64,
    pub duration: u64,
    #[serde(flatten)]
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub reference: HashMap<String, String>,
    pub service_name: String,
    #[serde(flatten)]
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub attributes: HashMap<String, json::Value>,
    #[serde(flatten)]
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub service: HashMap<String, json::Value>,
    pub events: String,
    pub links: String,
}

#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
pub enum SpanRefType {
    #[default]
    #[serde(rename = "CHILD_OF")]
    ChildOf,
    #[serde(rename = "PARENT_OF")]
    ParentOf,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
pub struct Event {
    pub name: String,
    pub _timestamp: u64,
    #[serde(flatten)]
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub attributes: HashMap<String, json::Value>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpanLinkContext {
    pub trace_id: String,
    pub span_id: String,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trace_flags: Option<u32>,
    #[serde(default)]
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trace_state: Option<String>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpanLink {
    pub context: SpanLinkContext,
    #[serde(flatten)]
    #[serde(skip_serializing_if = "HashMap::is_empty")]
    pub attributes: HashMap<String, json::Value>,
    #[serde(default)]
    pub dropped_attributes_count: u32,
}

#[derive(Debug, Default, Serialize, Deserialize, ToSchema)]
pub struct ExportTraceServiceResponse {
    // The details of a partially successful export request.
    //
    // If the request is only partially accepted
    // (i.e. when the server accepts only parts of the data and rejects the rest)
    // the server MUST initialize the `partial_success` field and MUST
    // set the `rejected_<signal>` with the number of items it rejected.
    //
    // Servers MAY also make use of the `partial_success` field to convey
    // warnings/suggestions to senders even when the request was fully accepted.
    // In such cases, the `rejected_<signal>` MUST have a value of `0` and
    // the `error_message` MUST be non-empty.
    //
    // A `partial_success` message with an empty value (rejected_<signal> = 0 and
    // `error_message` = "") is equivalent to it not being set/present. Senders
    // SHOULD interpret it the same way as in the full success case.
    pub partial_success: Option<ExportTracePartialSuccess>,
}

#[derive(Debug, Default, Serialize, Deserialize, ToSchema)]
pub struct ExportTracePartialSuccess {
    // The number of rejected spans.
    //
    // A `rejected_<signal>` field holding a `0` value indicates that the
    // request was fully accepted.
    pub rejected_spans: i64,

    // A developer-facing human-readable message in English. It should be used
    // either to explain why the server rejected parts of the data during a partial
    // success or to convey warnings/suggestions during a full success. The message
    // should offer guidance on how users can address such issues.
    //
    // error_message is an optional field. An error_message with an empty value
    // is equivalent to it not being set.
    pub error_message: String,
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use super::*;

    #[test]
    fn test_span() {
        let mut attributes = HashMap::new();
        attributes.insert("key".to_string(), json::json!("value"));

        let mut service = HashMap::new();
        service.insert("version".to_string(), json::json!("1.0"));

        let mut reference = HashMap::new();
        reference.insert("parent".to_string(), "parent-id".to_string());

        let span = Span {
            trace_id: "trace-1".to_string(),
            span_id: "span-1".to_string(),
            flags: 1,
            span_status: "OK".to_string(),
            span_kind: "SERVER".to_string(),
            operation_name: "test-operation".to_string(),
            start_time: 1000,
            end_time: 2000,
            duration: 1000,
            reference,
            service_name: "test-service".to_string(),
            attributes,
            service,
            events: "[]".to_string(),
            links: "[]".to_string(),
        };

        assert_eq!(span.trace_id, "trace-1");
        assert_eq!(span.span_id, "span-1");
        assert_eq!(span.flags, 1);
        assert_eq!(span.span_status, "OK");
        assert_eq!(span.span_kind, "SERVER");
        assert_eq!(span.operation_name, "test-operation");
        assert_eq!(span.start_time, 1000);
        assert_eq!(span.end_time, 2000);
        assert_eq!(span.duration, 1000);
        assert_eq!(span.service_name, "test-service");
        assert_eq!(span.events, "[]");
        assert_eq!(span.links, "[]");
    }

    #[test]
    fn test_event() {
        let mut attributes = HashMap::new();
        attributes.insert("key".to_string(), json::json!("value"));

        let event = Event {
            name: "test-event".to_string(),
            _timestamp: 1000,
            attributes,
        };

        assert_eq!(event.name, "test-event");
        assert_eq!(event._timestamp, 1000);
    }

    #[test]
    fn test_span_link_context() {
        let context = SpanLinkContext {
            trace_id: "trace-1".to_string(),
            span_id: "span-1".to_string(),
            trace_flags: Some(1),
            trace_state: Some("state".to_string()),
        };

        assert_eq!(context.trace_id, "trace-1");
        assert_eq!(context.span_id, "span-1");
        assert_eq!(context.trace_flags, Some(1));
        assert_eq!(context.trace_state, Some("state".to_string()));
    }

    #[test]
    fn test_span_link() {
        let mut attributes = HashMap::new();
        attributes.insert("key".to_string(), json::json!("value"));

        let link = SpanLink {
            context: SpanLinkContext {
                trace_id: "trace-1".to_string(),
                span_id: "span-1".to_string(),
                trace_flags: None,
                trace_state: None,
            },
            attributes,
            dropped_attributes_count: 0,
        };

        assert_eq!(link.context.trace_id, "trace-1");
        assert_eq!(link.context.span_id, "span-1");
        assert_eq!(link.dropped_attributes_count, 0);
    }

    #[test]
    fn test_export_trace_service_response() {
        let response = ExportTraceServiceResponse {
            partial_success: Some(ExportTracePartialSuccess {
                rejected_spans: 0,
                error_message: "".to_string(),
            }),
        };

        assert!(response.partial_success.is_some());
        let partial = response.partial_success.unwrap();
        assert_eq!(partial.rejected_spans, 0);
        assert_eq!(partial.error_message, "");
    }

    #[test]
    fn test_export_trace_partial_success() {
        let partial = ExportTracePartialSuccess {
            rejected_spans: 5,
            error_message: "test error".to_string(),
        };

        assert_eq!(partial.rejected_spans, 5);
        assert_eq!(partial.error_message, "test error");
    }
}

File: web/src/test/unit/mockData/traces.ts
// Copyright 2026 OpenObserve Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

// Formatted span node in the shape produced by TraceDetails getFormattedSpan,
// as it appears inside traceTree (children nested under `spans`).
const treeSpan = (
  spanId: string,
  serviceName: string,
  parentId: string,
  spans: any[],
  durationMs: number,
) => ({
  spanId,
  span_id: spanId,
  serviceName,
  resolvedIdentity: serviceName,
  parentId,
  durationMs,
  spans,
  operationName: `${serviceName}:op`,
  operation_name: `${serviceName}:op`,
  start_time: 1755853746625720300,
  spanStatus: "UNSET",
});

// Trace-tree fixtures for pattern-tree (Trace Graph) tests. Each fixture is
// the `traceTree` array TraceDetails builds — one entry per root span.
export const patternTraceTrees = {
  // Baseline: single root, root service calls one downstream service
  singleRoot: [
    treeSpan("a1", "alertmanager", "", [treeSpan("q1", "querier", "a1", [], 80)], 100),
  ],
  // Two root spans with distinct services, each with its own downstream call
  multiRootDistinctServices: [
    treeSpan("a1", "alertmanager", "", [treeSpan("q1", "querier", "a1", [], 80)], 100),
    treeSpan("i1", "ingester", "", [treeSpan("c1", "compactor", "i1", [], 40)], 60),
  ],
  // Two root spans that belong to the same service, no cross-service calls
  multiRootSameService: [
    treeSpan("a1", "alertmanager", "", [], 100),
    treeSpan("a2", "alertmanager", "", [], 50),
  ],
  // Two root spans whose services call each other (cyclic service relationship)
  multiRootCyclicServices: [
    treeSpan("a1", "alertmanager", "", [treeSpan("q1", "querier", "a1", [], 80)], 100),
    treeSpan("q2", "querier", "", [treeSpan("a2", "alertmanager", "q2", [], 30)], 50),
  ],
  // Root span calling another service, plus an orphan root of that same child service
  multiRootOrphanChildService: [
    treeSpan("a1", "alertmanager", "", [treeSpan("q1", "querier", "a1", [], 80)], 100),
    treeSpan("q2", "querier", "missing-parent", [], 30),
  ],
  // Single root span, single service, no relationships at all
  singleServiceOnly: [treeSpan("a1", "alertmanager", "", [], 100)],
};

export default {
  tracesDetails: {
    traceMeta: {
      hits: [
        [
          {
            duration: 295986,
            end_time: 1755853746921707300,
            first_event: {
              _timestamp: 1755853746625720,
              duration: 295986,
              end_time: 1755853746921707300,
              operation_name: "service:alerts:evaluate_scheduled",
              service_name: "alertmanager",
              span_status: "UNSET",
              start_time: 1755853746625720300,
              trace_id: "eab4575014a1fe101dba7de80a3cf6c3",
            },
            service_name: [
              {
                count: 4,
                service_name: "alertmanager",
              },
              {
                count: 31,
                service_name: "querier",
              },
              {
                count: 14,
                service_name: "ingester",
              },
            ],
            spans: [49, 0],
            start_time: 1755853746625720300,
            trace_id: "eab4575014a1fe101dba7de80a3cf6c3",
          },
        ],
      ],
    },
    traceSpans: {
      hits: [
        {
          _timestamp: 1755853746625720,
          busy_ns: "657334",
          code_filepath: "src/service/alerts/mod.rs",
          code_lineno: "114",
          code_namespace: "openobserve::service::alerts",
          duration: 295986,
          end_time: 1755853746921707300,
          events: "[]",
          flags: 1,
          idle_ns: "295329796",
          links: "[]",
          operation_name: "service:alerts:evaluate_scheduled",
          reference_parent_span_id: "d4b07e603e2fa32f",
          reference_parent_trace_id: "eab4575014a1fe101dba7de80a3cf6c3",
          reference_ref_type: "ChildOf",
          service_name: "alertmanager",
          service_service_instance: "dev2-openobserve-alertmanager-0",
          service_service_version: "v0.15.0-rc5",
          span_id: "6b080023171f5767",
          span_kind: "1",
          span_status: "UNSET",
          start_time: 1755853746625720300,
          status_code: 0,
          status_message: "",
          thread_id: "5",
          thread_name: "job_runtime",
          trace_id: "eab4575014a1fe101dba7de80a3cf6c3",
          _end_time_ns: "1755853746921707300",
          _start_time_ns: "1755853746625720300",
        },
        {
          _timestamp: 1755853746625779,
          busy_ns: "648454",
          code_filepath: "src/service/search/grpc_search.rs",
          code_lineno: "31",
          code_namespace: "openobserve::service::search::grpc_search",
          duration: 295923,
          end_time: 1755853746921702400,
          events: "[]",
          flags: 1,
          idle_ns: "295275675",
          links: "[]",
          operation_name: "service:search:grpc_search",
          reference_parent_span_id: "6b080023171f5767",
          reference_parent_trace_id: "eab4575014a1fe101dba7de80a3cf6c3",
          reference_ref_type: "ChildOf",
          service_name: "alertmanager",
          service_service_instance: "dev2-openobserve-alertmanager-0",
          service_service_version: "v0.15.0-rc5",
          span_id: "d427ced59acf399b",
          span_kind: "1",
          span_status: "UNSET",
          start_time: 1755853746625779000,
          status_code: 0,
          status_message: "",
          thread_id: "5",
          thread_name: "job_runtime",
          trace_id: "eab4575014a1fe101dba7de80a3cf6c3",
          _end_time_ns: "1755853746921702400",
          _start_time_ns: "1755853746625779000",
        },
        {
          _timestamp: 1755853746625790,
          busy_ns: "320271",
          code_filepath: "src/service/search/grpc_search.rs",
          code_lineno: "56",
          code_namespace: "openobserve::service::search::grpc_search",
          duration: 295178,
          end_time: 1755853746920969200,
          events: "[]",
          flags: 1,
          idle_ns: "294858953",
          links: "[]",
          node_addr: "http://10.1.105.53:5081",
          node_id: "5",
          operation_name: "service:search:cluster:grpc_search",
          reference_parent_span_id: "d427ced59acf399b",
          reference_parent_trace_id: "eab4575014a1fe101dba7de80a3cf6c3",
          reference_ref_type: "ChildOf",
          service_name: "alertmanager",
          service_service_instance: "dev2-openobserve-alertmanager-0",
          service_service_version: "v0.15.0-rc5",
          span_id: "bf6bde74cdcc245f",
          span_kind: "1",
          span_status: "UNSET",
          start_time: 1755853746625790200,
          status_code: 0,
          status_message: "",
          thread_id: "5",
          thread_name: "job_runtime",
          trace_id: "eab4575014a1fe101dba7de80a3cf6c3",
          _end_time_ns: "1755853746920969200",
          _start_time_ns: "1755853746625790200",
        },
      ],
    },
  },
};
